### PR TITLE
gh-104773: PEP 594: Remove the sndhdr module

### PR DESCRIPTION
**Pull request by vstinner**
Remove the Lib/test/sndhdrdata/ directory.

* Issue: gh-104773

----
:books: Documentation preview :books:: https://cpython-previews--104774.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
I ignored mentions of the sndhdr module in comments of the email module, Lib/email/mime/audio.py:

```py
# Originally from the sndhdr module.
#
# There are others in sndhdr that don't have MIME types. :(
# Additional ones to be added to sndhdr? midi, mp3, realaudio, wma??
def _what(data):
    # Try to identify a sound file type.
    #
    # sndhdr.what() had a pretty cruddy interface, unfortunately.  This is why
    # we re-do it here.  It would be easier to reverse engineer the Unix 'file'
    # command and use the standard 'magic' file, as shipped with a modern Unix.
    hdr = data[:512]
    fakefile = BytesIO(hdr)
    for testfn in _rules:
        if res := testfn(hdr, fakefile):
            return res
    else:
        return None
```

**Comment by vstinner** [comment]
Fix the Docs CI: ``Warning: (...)/Doc/whatsnew/3.12.rst:819: WARNING: py:mod reference target not found: sndhdr``.

**Review by gpshead — APPROVED** [review]

File: Doc/whatsnew/3.12.rst [hunk 1]
```diff
@@ -816,7 +816,7 @@ Modules (see :pep:`594`):
 * :mod:`nntplib`
 * :mod:`ossaudiodev`
 * :mod:`pipes`
-* :mod:`sndhdr`
+* ``sndhdr``
```
Comment (hugovk) [inline Doc/whatsnew/3.12.rst]: An alternative is to add an exclamation mark. This tells Sphinx not to look up the reference, but it still formats it consistently like a module.

```suggestion
* :mod:`!sndhdr`
```

File: Doc/whatsnew/3.13.rst [hunk 2]
```diff
@@ -102,6 +102,11 @@ Deprecated
 Removed
 =======
 
+* Remove the ``sndhdr`` module, deprecated in Python 3.11: use the projects
+  `filetype <https://pypi.org/project/filetype/>`_,
+  `puremagic <https://pypi.org/project/puremagic/>`_,
+  or `python-magic <https://pypi.org/project/python-magic/>`_ instead.
+  (Contributed by Victor Stinner in :gh:`104773`.)
```
Comment (hugovk) [inline Doc/whatsnew/3.13.rst:154]: Good idea to list the PyPI alternatives 👍 

Let's also add a reference to `` `:pep:`594` `` for people searching for it.

File: Misc/NEWS.d/next/Library/2023-05-23-01-37-40.gh-issue-104773.8c-GsG.rst
Comment (hugovk) [inline Misc/NEWS.d/next/Library/2023-05-23-01-37-40.gh-issue-104773.8c-GsG.rst:1]: Maybe also a `` `:pep:`594` `` reference here?

**Review by hugovk — COMMENTED** [review]
Please could you also update the refs in 3.11.rst?

For example with an exclamation mark:

```rst
* :pep:`594` led to the deprecations of the following modules
  slated for removal in Python 3.13:

  +---------------------+---------------------+---------------------+---------------------+---------------------+
  | :mod:`aifc`         | :mod:`chunk`        | :mod:`msilib`       | :mod:`pipes`        | :mod:`telnetlib`    |
  +---------------------+---------------------+---------------------+---------------------+---------------------+
  | :mod:`audioop`      | :mod:`crypt`        | :mod:`nis`          | :mod:`!sndhdr`      | :mod:`uu`           |
  +---------------------+---------------------+---------------------+---------------------+---------------------+
  | :mod:`cgi`          | :mod:`imghdr`       | :mod:`nntplib`      | :mod:`spwd`         | :mod:`xdrlib`       |
  +---------------------+---------------------+---------------------+---------------------+---------------------+
  | :mod:`cgitb`        | :mod:`mailcap`      | :mod:`ossaudiodev`  | :mod:`sunau`        |                     |
  +---------------------+---------------------+---------------------+---------------------+---------------------+

```

You could do them all in a single commit/PR, to avoid lots of conflicts later.

And:

```diff
diff --git a/Doc/whatsnew/3.5.rst b/Doc/whatsnew/3.5.rst
index 14b6425cea..a1b6e18950 100644
--- a/Doc/whatsnew/3.5.rst
+++ b/Doc/whatsnew/3.5.rst
@@ -1714,7 +1714,7 @@ support :rfc:`6531` (SMTPUTF8).
 sndhdr
 ------
 
-The :func:`~sndhdr.what` and :func:`~sndhdr.whathdr` functions  now return
+The ``what`` and ``whathdr`` functions now return
 a :func:`~collections.namedtuple`.  (Contributed by Claudiu Popa in
 :issue:`18615`.)
```

**Comment by vstinner** [comment]
I updated the doc: I added PEP 594 and replaced ``sndhdr`` with ``!sndhdr``. I also rebased the PR on the main to retrieve recent test fixes (especially test_importlib).

**Pull request merged**